benchmarks/Microsoft.AspNetCore.Mvc.Performance/MvcEndpointDatasourceBenchmark.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using BenchmarkDotNet.Attributes;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using Microsoft.AspNetCore.Mvc.ActionConstraints;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.AspNetCore.Mvc.Internal;
+using Microsoft.AspNetCore.Mvc.Routing;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Microsoft.AspNetCore.Mvc.Performance
+{
+    public class MvcEndpointDataSourceBenchmark
+    {
+        private const string DefaultRoute = "{Controller}/{Action=Index}/{id?}";
+
+        private MockActionDescriptorCollectionProvider _conventionalActionDescriptorCollectionProvider;
+        private MockActionDescriptorCollectionProvider _attributeActionDescriptorCollectionProvider;
+        private List<MvcEndpointInfo> _conventionalEndpointInfos;
+
+        [Params(1, 100, 1000)]
+        public int ActionCount;
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            _conventionalActionDescriptorCollectionProvider = new MockActionDescriptorCollectionProvider(
+                Enumerable.Range(0, ActionCount).Select(i => CreateActionDescriptor(i, false)).ToList()
+                );
+
+            _attributeActionDescriptorCollectionProvider = new MockActionDescriptorCollectionProvider(
+                Enumerable.Range(0, ActionCount).Select(i => CreateActionDescriptor(i, true)).ToList()
+                );
+
+            _conventionalEndpointInfos = new List<MvcEndpointInfo>
+            {
+                new MvcEndpointInfo(
+                    "Default",
+                    DefaultRoute,
+                    new RouteValueDictionary(),
+                    new Dictionary<string, object>(),
+                    new RouteValueDictionary(),
+                    new MockInlineConstraintResolver())
```

Comment:
It's really strange to see a service passed into a data object like this. (I'll be honest, I don't remember how we got here, and I don't expect to solve it in this PR).

Do you think this would be solved by using `RoutePattern` for more of this stuff in the future?

Reply:
Normally this object would be created via a builder

Reply:
it's still strange